### ISSUE TITLE
Stacks seem incorrect with offwaketime on Android

### ISSUE THREAD

**Issue by joelagnel** [iovisor/bcc#1679]
`offwaketime` is giving me some weird waker/sleep stacks on Android/ARM64.

Here's one of the stacks it found:
```
    waker:           find 0
    el0_svc_naked
    sys_newfstatat
    vfs_fstatat
    vfs_getattr_nosec
    kernfs_iop_getattr
    mutex_unlock
    __mutex_unlock_slowpath
    --               --
    try_to_wake_up
    kthread_queue_work
    sugov_irq_work
    irq_work_run_list
    irq_work_run
    handle_IPI
    gic_handle_irq
    el1_irq
    _raw_spin_unlock_irqrestore
    __wake_up_sync_key
    pipe_write
    new_sync_write
    vfs_write
    sys_write
    el0_svc_naked
    target:          ActivityManager 1163
        69
```
Basically the test was running `find /data | cksum` is running in the background with `offwaketime -K`.

What I see above is 3 unrelated stacks. ActivityManager isn't really involved in the find/cksum activity, yet its identified as a sleeper. Lets assume it did do a pipe_write but further down in the sleeper stack, it doesn't show schedule or any such call that can cause a sleep. All it shows is `try_to_wake_up` which is actually a wakeup function not a sleep function.

Also the `try_to_wake_up` is coming from an interrupt stack, and its a path that is handling an IPI related to cpufreq.  This IRQ stack isn't related to the pipe write or `vfs_fsstatat` at above. Its just pure IPI/cpufreq stuff.

Could it be that the IRQ stack (for the IPI) is messing up the offwaketime logic?

@brendangregg , how are we handling IRQ stacks with offwaketime? Like what happens if an IRQ wakes up a task while some other totally unrelated task is running? Is there a false waker/wakee link established there somehow? Any other ideas on the above stack? Thanks!

**Comment by yonghong-song** [comment]
On Thu, Apr 12, 2018 at 1:18 PM, Joel <notifications@github.com> wrote:
> offwaketime is giving me some weird waker/sleep stacks on Android/ARM64.
>
> Here's one of the stacks it found:
>
>     waker:           find 0
>     el0_svc_naked
>     sys_newfstatat
>     vfs_fstatat
>     vfs_getattr_nosec
>     kernfs_iop_getattr
>     mutex_unlock
>     __mutex_unlock_slowpath
>     --               --
>     try_to_wake_up
>     kthread_queue_work
>     sugov_irq_work
>     irq_work_run_list
>     irq_work_run
>     handle_IPI
>     gic_handle_irq
>     el1_irq
>     _raw_spin_unlock_irqrestore
>     __wake_up_sync_key
>     pipe_write
>     new_sync_write
>     vfs_write
>     sys_write
>     el0_svc_naked
>     target:          ActivityManager 1163
>         69
>
> Basically the test was running find /data | cksum is running in the
> background with offwaketime -K.
>
> What I see above is 3 unrelated stacks. ActivityManager isn't really
> involved in the find/cksum activity, yet its identified as a sleeper. Lets
> assume it did do a pipe_write but further down in the sleeper stack, it
> doesn't show schedule or any such call that can cause a sleep. All it shows
> is try_to_wake_up which is actually a wakeup function not a sleep function.
>
> Also the try_to_wake_up is coming from an interrupt stack, and its a path
> that is handling an IPI related to cpufreq. This IRQ stack isn't related to
> the pipe write or vfs_fsstatat at above. Its just pure IPI/cpufreq stuff.
>
> Could it be that the IRQ stack (for the IPI) is messing up the offwaketime
> logic?

In this case, IPI wakes up the task. The ActivityManager is involved because
it is interrupted and it is part of kernel stack. One way to avoid such cases
may be filtering based on known kernel irq-related functions. For example,
if we see handle_IPI is part of kernel stack, we are certain that IPI
is involved
and probably this waker-wakee relationship is not worth reporting by default, or
reporting by default but having an option not to report.

Each individual stack reporting, even with IPI, should not be an issue in
my opinion since it is quite clear that IPI caused the task wakeup.
The issue may be in aggregation where people may get confused if legitimate
and non-legitimate stacks mixed together.

**Comment by joelagnel** [comment]
Yes, instead of only `handle_IPI`, I think its better to just not track any IRQ related wake ups at all. We ran into a similar issue before and one of the ideas floating around was to look at the IRQ flags at the time the hook runs. I am not fully sure if we can access that from the BPF program. I will try to investigate such solutions. I happen to be geographically close to Peterz and my BPF talk is tomorrow so I will bring this problem up there as well :)

**Comment by joelagnel** [comment]
Another idea could be:
1. only unwind IRQ stack if its an IRQ wake up
2. make waker task IRQ number or something, instead of a victim task

**Comment by yonghong-song** [comment]
@joelagnel What kind of irq information on which data structure do we need to identify it is a irq context? If it is available from `current` (recursively), we can get it from inside bpf program.

What does "only unwind IRQ stack if it is an IRQ wake up"? You mean a task is sleeping with intention to be woken up by an IRQ? Do you need woken-up process information to confirm this?

**Comment by palmtenor** [comment]
@joelagnel , can you give it another try after #1693 and #1704?

**Comment by joelagnel** [comment]
Yes, sure. I was out on conference travel so bear with me as I deal with my Inbox avalanche. Will try out the mentioned PRs today. Thanks @yonghong-song and @palmtenor for looking into this.

**Comment by joelagnel** [comment]
@yonghong-song , Re: "only unwind IRQ stack if it is an IRQ wake up".
I meant, if the wake up is because of a process, then unwind only the process stack, not the IRQ stack. So for a case like this:
```
process 1 -> syscall -> driver -> wake up -> process 2
```
On the other hand, if it was an IRQ driven wake up, like a timer callback.
```
# process 1 is running and doing something unrelated
# Now,
irq -> wake up -> process 2
```
In this case, it doesn't make sense to unwind the process 1 stack, so we should only unwind from the IRQ stack. I remember this happening, but I will test for this case again and make sure its really unwinding an unrelated stack.

**Comment by joelagnel** [comment]
@yonghong-song Re: "What kind of irq information on which data structure do we need to identify it is a irq context?"
The IRQ context information should be available in the preempt flags, by looking at the HARDIRQ offset in those flags (or something like that). I'm not sure if this is made available to tracepoint/kprobe callbacks as part of the arguments. It could probably be done through a helper.

**Comment by joelagnel** [comment]
@palmtenor I think the situation has improved with those 2 PRs (I am not sure because I'm unable to see the exact same stack with bcc master). However, the wrong IRQ waker issue remains:
```
    waker:           Binder:1167_1E 9280
    el0_irq_naked
    gic_handle_irq
    __handle_domain_irq
    handle_fasteoi_irq
    handle_irq_event
    __handle_irq_event_percpu
    ufshcd_intr
    ufshcd_transfer_req_compl
    __ufshcd_transfer_req_compl
    ufs_qcom_pm_qos_req_end
    queue_work_on
    __queue_work
    --               --
    finish_task_switch
    preempt_schedule_common
    __sched_text_start
    _raw_spin_unlock_irq
    process_one_work
    worker_thread
    kthread
    ret_from_fork
    target:          kworker/u16:13 3311
        16
```
In the above, the kworker, is woken up by an interrupt (probably for block I/O completion). However it appears that binder is the waker. Binder just happens to be the task running at the IRQ interruption time. Its not really the waker.

**Comment by yonghong-song** [comment]
>  I'm not sure if this is made available to tracepoint/kprobe callbacks as part of the arguments. It could probably be done through a helper.

Such information is not available to bpf program at this time, but a helper could be added to get such information.

**Comment by joelagnel** [comment]
Ok I'll work on something like this and try to make use of it from this tracer. It could be useful to other usecases as well in the future.

One question: If IRQ context caused the wake, how do we make the tool show output? Do we discard such wake ups, or annotate them somehow with this information?

**Comment by yonghong-song** [comment]
I think we probably still want to show them (waken up by IRQ context) but we may not want to show the interrupted process information since they are irrelevant?
